src/common/LinkParser.cpp

```diff
@@ -1,206 +1,231 @@
+#define QT_NO_CAST_FROM_ASCII  // avoids unexpected implicit casts
 #include "common/LinkParser.hpp"
 
 #include <QFile>
-#include <QMap>
-#include <QRegularExpression>
 #include <QSet>
 #include <QString>
-#include <QStringRef>
+#include <QStringView>
 #include <QTextStream>
 
-namespace chatterino {
 namespace {
-    QSet<QString> &tlds()
-    {
-        static QSet<QString> tlds = [] {
-            QFile file(":/tlds.txt");
-            file.open(QFile::ReadOnly);
-            QTextStream stream(&file);
+
+QSet<QString> &tlds()
+{
+    static QSet<QString> tlds = [] {
+        QFile file(QStringLiteral(":/tlds.txt"));
+        file.open(QFile::ReadOnly);
+        QTextStream stream(&file);
 
 #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
-            // Default encoding of QTextStream is already UTF-8, at least in Qt6
+        // Default encoding of QTextStream is already UTF-8, at least in Qt6
 #else
-            stream.setCodec("UTF-8");
+        stream.setCodec("UTF-8");
 #endif
-            int safetyMax = 20000;
+        int safetyMax = 20000;
 
-            QSet<QString> set;
+        QSet<QString> set;
 
-            while (!stream.atEnd())
-            {
-                auto line = stream.readLine();
-                set.insert(line);
+        while (!stream.atEnd())
+        {
+            auto line = stream.readLine();
+            set.insert(line);
 
-                if (safetyMax-- == 0)
-                    break;
+            if (safetyMax-- == 0)
+            {
+                break;
             }
+        }
 
-            return set;
-        }();
-        return tlds;
-    }
+        return set;
+    }();
+    return tlds;
+}
 
-    bool isValidHostname(QStringRef &host)
-    {
-        int index = host.lastIndexOf('.');
+bool isValidTld(QStringView tld)
+{
+    return tlds().contains(tld.toString().toLower());
+}
 
-        return index != -1 &&
-               tlds().contains(host.mid(index + 1).toString().toLower());
-    }
+bool isValidIpv4(QStringView host)
+{
+    // We don't care about the actual value,
+    // we only want to verify the ip.
 
-    bool isValidIpv4(QStringRef &host)
+    char16_t sectionValue = 0;  // 0..256
+    uint8_t octetNumber = 0;    // 0..4
+    uint8_t sectionDigits = 0;  // 0..3
+    bool lastWasDot = true;
+
+    for (auto i : host)
     {
-        static auto exp = QRegularExpression("^\\d{1,3}(?:\\.\\d{1,3}){3}$");
+        char16_t current = i.unicode();
+        if (current == '.')
+        {
+            if (lastWasDot || octetNumber == 3)
+            {
+                return false;
+            }
+            lastWasDot = true;
+            octetNumber++;
+            sectionValue = 0;
+            sectionDigits = 0;
+            continue;
+        }
+        lastWasDot = false;
 
-        return exp.match(host).hasMatch();
+        if (current > u'9' || current < u'0')
```

Comment:
[QChar::isDigit](https://doc.qt.io/qt-6/qchar.html#isDigit) implementation seems to handle unicode stuff, your solution seems good to me

Reply:
Right, `isDigit` also checks if the character is the "Nd" unicode category.